### ISSUE TITLE
Require pull requests to rebase on top of master prior to being merged

### ISSUE THREAD

**Issue by dhil** [links-lang/links#396]
We use the squash'n'merge strategy for merging in pull requests on the master branch. This is good, because it allows us to develop our changes incrementally in many small commits that not necessarily meaningful on their own, but when merged they get squashed into one big meaningful commit.

There is however a "hygiene problem" with the squash'n'merge strategy as it makes possible to re-merge changes that are already present in the master branch. This is for example happening in #385 PR which merges some +5000 changes of which several hundreds are redundant (I don't have anything against this particular PR being merged as-is; it just reminded me of the problem).

The redundancy does not cause any harm to the master branch per see, but it does have irksome side-effects such as increasing the likelihood of merge-conflicts and obfuscating the commit history. 

For example, say that branch A introduces a redundant commit into the master branch and B is another active branch. Suppose that B was recently rebased on top of the master branch during which some merge conflicts were resolved. Merging A into master may reintroduce the merge conflicts next time B rebases on top of master. 

Moreover, the author of the redundant commit gets credited for the work, meaning that -- as far as git is concerned -- the author takes ownership of the code. This can prove to be problematic when trying to nail down who-wrote-what.

Thus I suggest that we require as a minimum that every pull request must be able to be rebased on the master branch before it is merged in. I reckon we can even instrument the continuous integration tool to check this invariant. Adding something along the lines of the following command to the `script` section in the Travis configuration file ought to work.
```
script:
- git fetch && git rebase origin/master
```

In summary: I propose we continue to use squash'n'merge as our primary strategy, but we require that any pull request is at least able to rebase on top of the current master as well.
